zerver/views/state.py

```diff
@@ -0,0 +1,47 @@
+from django.http import HttpRequest, HttpResponse
+from django.utils.translation import ugettext as _
+from zerver.lib.bot_storage import (
+    get_bot_state,
+    set_bot_state,
+    remove_bot_state,
+    get_keys_in_bot_state,
+    is_key_in_bot_state,
+    StateError,
+)
+from zerver.decorator import has_request_variables, REQ
+from zerver.lib.response import json_success, json_error
+from zerver.lib.validator import check_dict, check_list, check_string
+from zerver.models import UserProfile
+
+from typing import Dict, List, Optional
+
+@has_request_variables
+def update_state(request, user_profile, state=REQ(validator=check_dict([]))):
+    # type: (HttpRequest, UserProfile, Optional[Dict[str, str]]) -> HttpResponse
+    try:
+        for key, value in state.items():
+            set_bot_state(user_profile, key, value)
+        return json_success()
+    except StateError as e:
+        return json_error(str(e))
```

Comment:
These feel a bit choppy in how they're written; I'd consider moving the `except StateError` block to just be around the `set_bot_state` line, not the whole function.  Generally, you want except blocks to be as narrowed scoped as is correct for maximum readability.